torch/_functorch/aot_autograd.py

```diff
@@ -508,14 +508,16 @@ class SubclassCreationMeta:
     #  so holding onto this at runtime shouldn't leak memory)
     original_subclass: torch.Tensor
     # meta and inner_keys are produced by the subclass's __tensor_flatten__.
-    # We need to keep them around to plumb them into __tensor_unflatten__.
+    # We need to keep them around along with outer_size to plumb them into __tensor_unflatten__.
     meta: Any
     inner_keys: List[any]
+    outer_size: Tuple[int]
 
     def creation_fn(self, all_args, *, is_runtime: bool):
         curr_args = all_args[self.flat_tensor_start_idx:self.flat_tensor_start_idx + self.arg_count]
         assert len(curr_args) == len(self.inner_keys), f'inner_keys: {str(self.inner_keys)}. len(curr_args): {len(curr_args)}'
-        out = type(self.original_subclass).__tensor_unflatten__(dict(zip(self.inner_keys, curr_args)), self.meta)
+        out = type(self.original_subclass).__tensor_unflatten__(
+            dict(zip(self.inner_keys, curr_args)), self.meta, self.outer_size)
```

Comment:
Hmm, I wonder if we'll have a problem here later.

`self.outer_size` here will be a `List[SymInt]`, since we stashed it at compile time, when we first traced the (faketensor) output from tracing the forward).

But this function is also used at runtime, to create the runtime tensor subclass to the user, after the compiler hands us dense tensor outputs.

Reply:
I can think of two options:

(1) Make it the responsibility of the subclass's `__tensor_unflatten__`: if it sees that its being passed real inner tensors, but symbolic outer sizes, it should know to discard the outer sizes. Downsides: (a) it's ugly and puts more burden on subclasses, (b) we might have some subclasses where they have no way of actually computing the outer size if we don't pass it one

(2) Use the shape-env, plus the runtime input shapes of the graph, to figure out what the runtime output shapes are, and plumb real shapes here. I think this feels like the right thing to do, but I'm not sure how easy it is.

Reply:
from offline discussion: I'm de facto doing (1) but punting on making this nicer to future work